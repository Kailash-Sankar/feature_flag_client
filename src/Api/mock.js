function getPackages() {
  return {
    "0": { id: 0, name: "Custom" },
    "1": { id: 1, name: "Silver" },
    "2": { id: 2, name: "Gold" },
    "3": { id: 3, name: "Diamond" }
  };
}

// hardcoded for demo
function getPackageFeatures() {
  return {
    "0": [],
    "1": {
      ad_slates: {
        id: "ad_slates",
        status: 2
      }
    },
    "2": [],
    "3": []
  };
}

export default {
  getPackages,
  getPackageFeatures
};
