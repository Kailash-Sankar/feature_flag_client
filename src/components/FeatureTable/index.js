import React from "react";
import { Collapse, Radio } from "antd";
const { Panel } = Collapse;
import * as styles from "./index.module.less";

const text = <p style={{ paddingLeft: 24 }}>Feature realted attributes</p>;

const statusMap = {
  "0": "Disabled",
  "1": "Internal",
  "2": "All"
};

export function StatusSelector({ fid, value, onChange }) {
  return (
    <Radio.Group onChange={onChange} value={value}>
      <Radio value={0} data-id={fid}>
        Disabled
      </Radio>
      <Radio value={1} data-id={fid}>
        Internal
      </Radio>
      <Radio value={2} data-id={fid}>
        All
      </Radio>
    </Radio.Group>
  );
}

export function FeatureBar({ feature, pos, onUpdate, readOnly }) {
  function onChange(e) {
    const fid = e.target["data-id"];
    const value = e.target.value;
    onUpdate(fid, pos, value);
  }

  const handleClick = (event) => event.stopPropagation();

  return (
    <div onClick={handleClick}>
      {readOnly ? (
        <span>{statusMap[feature.status]}</span>
      ) : (
        <StatusSelector
          value={feature.status}
          onChange={onChange}
          fid={feature.id}
        />
      )}
    </div>
  );
}

function FeatureTable({ features, updateFeatures, readOnly = false }) {
  return (
    <Collapse bordered={false} defaultActiveKey={[]}>
      {features.map((f, index) => {
        return (
          <Panel
            header={<RenderHeader data={f} />}
            extra={
              <FeatureBar
                feature={f}
                pos={index}
                onUpdate={updateFeatures}
                readOnly={readOnly}
              />
            }
            key={f.id}
          >
            {text}
          </Panel>
        );
      })}
    </Collapse>
  );
}

export function RenderHeader({ data }) {
  return (
    <div style={{ display: "inline-block" }}>
      <div>{data.name}</div>
      <div className={styles.sub}>{data.description}</div>
    </div>
  );
}

export default FeatureTable;
