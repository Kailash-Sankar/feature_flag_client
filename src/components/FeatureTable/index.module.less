.sub {
  font-size: 12px;
  color: #757575;
}
