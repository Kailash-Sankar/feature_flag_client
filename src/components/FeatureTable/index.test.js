import { shallow } from "enzyme";
import React from "react";
import mock from "@mock/data";

import FeatureTable, {
  StatusSelector,
  FeatureBar,
  RenderHeader
} from "./index";

describe("Feature Table", () => {
  test("status selector", () => {
    const wrapper = shallow(
      <StatusSelector fid="ff_one" value={0} onChange={mock.noop} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test("feature bar editable", () => {
    const wrapper = shallow(
      <FeatureBar
        feature={mock.features[0]}
        pos={0}
        onUpdate={mock.noop}
        readOnly={false}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test("feature bar readonly", () => {
    const wrapper = shallow(
      <FeatureBar
        feature={mock.features[0]}
        pos={0}
        onUpdate={mock.noop}
        readOnly={true}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test("render header", () => {
    const wrapper = shallow(<RenderHeader data={mock.features[0]} />);
    expect(wrapper).toMatchSnapshot();
  });

  test("table", () => {
    const props = {
      features: mock.features,
      updateFeatures: mock.noop,
      readOnly: false
    };
    const wrapper = shallow(<FeatureTable {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
