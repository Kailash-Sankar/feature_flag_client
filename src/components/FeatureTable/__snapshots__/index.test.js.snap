// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feature Table feature bar editable 1`] = `
<div
  onClick={[Function]}
>
  <StatusSelector
    fid="ff_one"
    onChange={[Function]}
    value={2}
  />
</div>
`;

exports[`Feature Table feature bar readonly 1`] = `
<div
  onClick={[Function]}
>
  <span>
    All
  </span>
</div>
`;

exports[`Feature Table render header 1`] = `
<div
  style={
    Object {
      "display": "inline-block",
    }
  }
>
  <div>
    Feature One
  </div>
  <div>
    Enable feature to do something
  </div>
</div>
`;

exports[`Feature Table status selector 1`] = `
<RadioGroup
  buttonStyle="outline"
  onChange={[Function]}
  value={0}
>
  <Radio
    data-id="ff_one"
    type="radio"
    value={0}
  >
    Disabled
  </Radio>
  <Radio
    data-id="ff_one"
    type="radio"
    value={1}
  >
    Internal
  </Radio>
  <Radio
    data-id="ff_one"
    type="radio"
    value={2}
  >
    All
  </Radio>
</RadioGroup>
`;

exports[`Feature Table table 1`] = `
<Collapse
  bordered={false}
  defaultActiveKey={Array []}
  expandIconPosition="left"
>
  <CollapsePanel
    extra={
      <FeatureBar
        feature={
          Object {
            "attributes": Array [],
            "description": "Enable feature to do something",
            "id": "ff_one",
            "name": "Feature One",
            "product": "ax",
            "status": 2,
          }
        }
        onUpdate={[Function]}
        pos={0}
        readOnly={false}
      />
    }
    header={
      <RenderHeader
        data={
          Object {
            "attributes": Array [],
            "description": "Enable feature to do something",
            "id": "ff_one",
            "name": "Feature One",
            "product": "ax",
            "status": 2,
          }
        }
      />
    }
    key="ff_one"
  >
    <p
      style={
        Object {
          "paddingLeft": 24,
        }
      }
    >
      Feature realted attributes
    </p>
  </CollapsePanel>
  <CollapsePanel
    extra={
      <FeatureBar
        feature={
          Object {
            "attributes": Array [],
            "description": "Enable feature to do something",
            "id": "ff_two",
            "name": "Feature Two",
            "product": "ax",
            "status": 1,
          }
        }
        onUpdate={[Function]}
        pos={1}
        readOnly={false}
      />
    }
    header={
      <RenderHeader
        data={
          Object {
            "attributes": Array [],
            "description": "Enable feature to do something",
            "id": "ff_two",
            "name": "Feature Two",
            "product": "ax",
            "status": 1,
          }
        }
      />
    }
    key="ff_two"
  >
    <p
      style={
        Object {
          "paddingLeft": 24,
        }
      }
    >
      Feature realted attributes
    </p>
  </CollapsePanel>
  <CollapsePanel
    extra={
      <FeatureBar
        feature={
          Object {
            "attributes": Array [],
            "description": "Enable feature to do something",
            "id": "ff_three",
            "name": "Feature Three",
            "product": "cx",
            "status": 0,
          }
        }
        onUpdate={[Function]}
        pos={2}
        readOnly={false}
      />
    }
    header={
      <RenderHeader
        data={
          Object {
            "attributes": Array [],
            "description": "Enable feature to do something",
            "id": "ff_three",
            "name": "Feature Three",
            "product": "cx",
            "status": 0,
          }
        }
      />
    }
    key="ff_three"
  >
    <p
      style={
        Object {
          "paddingLeft": 24,
        }
      }
    >
      Feature realted attributes
    </p>
  </CollapsePanel>
</Collapse>
`;
